src/test/java/org/alfresco/heartbeat/HBDataCollectorServiceImplTest.java

```diff
@@ -52,76 +61,168 @@
     public void setUp()
     {
         mockDataSenderService = mock(HBDataSenderService.class);
-        dataCollectorService = new HBDataCollectorServiceImpl(true);
+
+        dataCollectorService = spy(new HBDataCollectorServiceImpl(true));
         dataCollectorService.setHbDataSenderService(mockDataSenderService);
     }
 
-
-
     @Test
-    public void testInitialEnabledEqualsDefaultState()
+    public void testInitialEnabledEqualsDefaultState() throws Exception
     {
-        HBDataCollectorService dataCollectorService = new HBDataCollectorServiceImpl(true);
+        HBDataCollectorServiceImpl dataCollectorService = new HBDataCollectorServiceImpl(true);
         assertTrue(dataCollectorService.isEnabledByDefault());
 
         dataCollectorService = new HBDataCollectorServiceImpl(false);
         assertFalse(dataCollectorService.isEnabledByDefault());
     }
 
+    private void activateHB(HBDataCollectorServiceImpl dataCollectorService, boolean activate)
+    {
+        LicenseDescriptor mockLicenseDescriptor = mock(LicenseDescriptor.class);
+        when(mockLicenseDescriptor.isHeartBeatDisabled()).thenReturn(!activate);
+        dataCollectorService.onLicenseChange(mockLicenseDescriptor);
+    }
+
+    @Test
+    public void testWrongCronExpression() throws Exception
+    {
+        HBDataCollectorServiceImpl dataCollectorService = new HBDataCollectorServiceImpl(true);
+
+        activateHB(dataCollectorService, true);
+
+        Scheduler scheduler = spy(StdSchedulerFactory.getDefaultScheduler());
+        dataCollectorService.setScheduler(scheduler);
+
+        // try with correct cron expression
+        SimpleHBDataCollector collector1 = new SimpleHBDataCollector("collector1");
+        collector1.setCronExpression("0 0 0/1 * * ?");
+
+        dataCollectorService.registerCollector(collector1);
+
+        verifyNumbersOfUnschedulesAndSchedules(scheduler, 1, 1);
+
+        // try with wrong cron expression
+        SimpleHBDataCollector collector2 = new SimpleHBDataCollector("collector2");
+        collector2.setCronExpression("0 0 0/1 wrong cron expr ?");
+
+        dataCollectorService.registerCollector(collector2);
+
+        verifyNumbersOfUnschedulesAndSchedules(scheduler, 2, 1);
+    }
 
     @Test
-    public void testHBDataSenderServiceEnabledChange()
+    public void testSchedulingWithDefaultEnabled() throws Exception
     {
```

Comment:
It looks like you covered the new functionality here which is good. However, maybe it is just a matter of taste but you could split this test up into its respective components. For example, if you have a look at my old HeartBeatTest you will see that each scenario is in its own test method. That way if one of the scenarios fails we will know what is going wrong and it is easy to see that each scenario is covered. The new HBDataCollectorServiceImpl should pass all of the old tests plus more.

Reply:
Just for reference, here are the scenarios covered previously :
* test that collectors get scheduled when HB is enabled by default - not through license
* test that collectors **don't** get scheduled when HB is disabled by default - not through license
* test that collectors are not scheduled when HB is disabled by default **but** all are scheduled once enabled in license
* test that collectors are scheduled when HB is enabled by default **but** all are unscheduled once disabled in license
* test that license fail reverts back to default enabled/disabled state and all collectors are scheduled/unscheduled

The one difference is that before the tests checked a single HB job and now we need to check jobs for all collectors